tests/phpunit/Reflection/ContainerReflection.php

```diff
@@ -0,0 +1,153 @@
+<?php
+/**
+ * This code is licensed under the BSD 3-Clause License.
+ *
+ * Copyright (c) 2017, Maks Rafalko
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\Tests\Reflection;
+
+use Closure;
+use Error;
+use Infection\Container;
+use ReflectionClass;
+use ReflectionMethod;
+use ReflectionNamedType;
+use ReflectionProperty;
+use function str_starts_with;
+use Webmozart\Assert\InvalidArgumentException as AssertException;
+
+final class ContainerReflection
+{
+    /** @var ReflectionClass<Container> */
+    private readonly ReflectionClass $reflection;
+
+    private readonly Closure $createServiceClosure;
+
+    private readonly Closure $getServiceClosure;
+
+    private readonly ReflectionProperty $factories;
+
+    private readonly ReflectionProperty $values;
+
+    public function __construct(private readonly Container $container)
+    {
+        $this->reflection = new ReflectionClass($container);
+
+        $this->factories = $this->reflection->getProperty('factories');
+        $this->values = $this->reflection->getProperty('values');
+
+        $this->createServiceClosure = $this->reflection->getMethod('createService')->getClosure($container);
+        $this->getServiceClosure = $this->reflection->getMethod('get')->getClosure($container);
```

Comment:
I suppose we can make these methods public, and add more methods to the container itself.